### DISCUSSION TITLE
New to FP, trying to work with Ramda in a TypeScript React app

### DISCUSSION THREAD

**Discussion by joshfarrant** [ramda/ramda#3138] [Q&A]
Hey all :wave:

I'm trying to get my head around how to work with Ramda in the context of a React app. For those familiar with React, this is a simplified version of how I would usually hide/show a menu.

```jsx
const MyComponent = () => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <div>
      <button
        onClick={() => {
          setIsOpen(prevState => !prevState);
        }}
      >
        Toggle
      </button>
      {isOpen && <h1>I'm open!</h1>}
    </div>
  );
};
```

I'm trying to figure out how to refactor this to use Ramda, mainly as a learning exercise.

I originally started trying to do something like this, although this feels like a bit of a mess.

```jsx
const MyComponent = () => {
  const toggle = R.complement(Boolean);
  const [isOpen, setIsOpen] = useState(false);

  return (
    <div>
      <button
        onClick={() => {
          setIsOpen(toggle);
        }}
      >
        Toggle
      </button>
      {isOpen && <h1>I'm open!</h1>}
    </div>
  );
};
```

I then started thinking about doing something like this, which _maybe_ feels a bit nicer, but there's a distinct lack of Ramda!

```jsx
const toggleState = stateSetter => () => stateSetter(x => !x);

const MyComponent = () => {
  const [isOpen, setIsOpen] = useState(false);
  const toggleIsOpen = toggleState(setIsOpen);

  return (
    <div>
      <button
        onClick={toggleIsOpen}
      >
        Toggle
      </button>
      {isOpen && <h1>I'm open!</h1>}
    </div>
  );
};
```

How would those more experienced with using Ramda within the context of a React app approach this? 🙂

**Comment by voodoohop** [comment]
The cleanest functional way in my opinion is to create a useToggle hook, which you can then reuse.



```javascript

const useToggle=(initalState = false) => {
  const [toggleState,setToggle] = useState(initialState);
  const toggle = () => setToggle(!toggleState); // could use R.not but its overkill
  return [toggleState, toggle];
}

const MyComponent = () => {
  const [isOpen, toggleOpen] = useToggle();

  return (
    <div>
      <button
        onClick={toggleOpen)
      >
        Toggle
      </button>
      {isOpen && <h1>I'm open!</h1>}
    </div>
  );
};
```

It may also be possible with some other Ramda function that could wrap the `setIsOpen` call. You could use a useCallback to cache the toggling function too but maybe it's unnecessary in this simple example.

**Reply by joshfarrant** [reply]
Yea that's a clean way of doing it, perhaps I'm trying to shoehorn Ramda in at this point!

**Reply by joshfarrant** [reply]
I'd also managed to miss `R.not`, so thanks for that too

**Reply by voodoohop** [reply]
https://tommmyy.github.io/ramda-react-redux-patterns/pages/react-ramda.html

**Reply by voodoohop** [reply]
It could be done like this with Ramda but doesn't look as readable to me.

```javascript
const useToggle = (initalState = false) => {
  const [toggleState, setToggle] = useState(initialState);
  const toggle = R.pipe(R.always(R.not(toggleState)), setToggle);
  return [toggleState, toggle];
}
```

**Comment by CrossEye** [comment]
Please keep in mind that Ramda is a library and not a framework.  It's designed for one key purpose: to make it simpler to write in JS the kind of code that's normal to write in functional languages.  So generally, I would be wary about trying to convert anything to use Ramda.
You point out that this is a learning exercise, and there's more room for such changes in that case, but I see nothing in the original code that would make me ask, "How can I do this in a more functional manner?"  So I would see little reason to introduce Ramda into this.

Ramda is not a general-purpose library in the tradition of Underscore and lodash.  It shines best in data-transformation, where it helps makes code much more declarative.  I do use it in UIs, but in a much more limited fashion.  I don't think it offers much here.

**Reply by dsmileybrainly** [reply]
Thank you for your reply! We love ramda at Brainly. We use ramda a lot in data transformation. Would you consider code generators a form of data transformation? We use ramda extensively in our Angular Schematics which generate code for our React applications.

**Reply by CrossEye** [reply]
@dsmileybrainly:

Absolutely.  In fact last year I built a fairly large code-generation system (in-house, not publicly available) that used Ramda extensively, alongside much nasty string manipulation.  Ramda definitely made the code cleaner.

It's not so much that there is one great use-case for Ramda.  There are plenty of places where it can be useful.  But as Ramda doesn't play well with stateful systems, UIs are not usually one of the better fits.  However I've certainly used it there too.
